File: jas_ocaml/test/tools/yaml_tool_line_test.ml
(** Line-tool gesture-seam tests — OCaml port of the Rust line seam
    tests in jas_dioxus/src/tools/yaml_tool.rs (the line_parity_*
    family). Structurally modeled on yaml_tool_blob_brush_test.ml:
    same seam, same loader pattern, same empty-layer model, same
    tool-state accessor for mode.

    These drive the PRODUCTION line tool loaded from the workspace
    bundle through on_press / on_move / on_release and assert the
    committed Element.Line.

    The line tool is simple: a press-drag-release commits a single
    Line whose endpoints are the press point (x1,y1) and the release
    point (x2,y2) in doc space; a drag shorter than 2pt (hypot) is
    rejected. It reads NO app-level state, so unlike the blob-brush
    seam tests there is NO bridge / seed call here.

    Seam mapping from Rust to OCaml:
      on_press     -> on_press ctx x y ~shift ~alt
      on_move(drag)-> on_move ctx x y ~shift ~alt ~dragging
      on_release   -> on_release ctx x y ~shift ~alt
      tool_state("mode") -> tool#tool_state "mode" (a Yojson string) *)

open Jas

let () = ignore (GMain.init ())

let line_tool () : Yaml_tool.yaml_tool option =
  match Workspace_loader.load () with
  | None -> None
  | Some ws ->
    match Workspace_loader.json_member "tools" ws.data with
    | Some (`Assoc tools) ->
      (match List.assoc_opt "line" tools with
       | Some spec -> Yaml_tool.from_workspace_tool spec
       | None -> None)
    | _ -> None

(** Document with a single empty layer (no children). *)
let empty_layer_model () : Model.model =
  let layer = Element.Layer {
    name = Some "L";
    id = None;
    children = [||];
    transform = None; locked = false; opacity = 1.0;
    visibility = Preview; blend_mode = Normal; mask = None;
    isolated_blending = false; knockout_group = false;
  } in
  let m = Model.create () in
  m#set_document_unbracketed (Document.make_document [| layer |]);
  m

let make_ctx model =
  let ctrl = new Controller.controller ~model () in
  let ctx : Canvas_tool.tool_context = {
    model;
    controller = ctrl;
    hit_test_selection = (fun _ _ -> false);
    hit_test_handle = (fun _ _ -> None);
    hit_test_text = (fun _ _ -> None);
    hit_test_path_curve = (fun _ _ -> None);
    request_update = (fun () -> ());
    draw_element_overlay = (fun _cr _elem ~is_partial:_ _cps -> ());
  } in
  (ctx, ctrl)

(** Children of the first layer. *)
let layer0_children (m : Model.model) : Element.element array =
  match m#document.layers.(0) with
  | Element.Layer { children; _ } -> children
  | _ -> [||]

(* The mode tool-state, as a plain string. *)
let mode (tool : Yaml_tool.yaml_tool) : string =
  match tool#tool_state "mode" with
  | `String s -> s
  | other -> Yojson.Safe.to_string other

(* ── Loader sanity ─────────────────────────────────── *)

let load_tests = [
  Alcotest.test_case "line_tool_loads_from_workspace" `Quick (fun () ->
    match line_tool () with
    | Some tool ->
      assert (tool#spec.id = "line")
    | None -> Alcotest.skip ());
]

(* ── draw_line commits one Line with the press/release endpoints ── *)

let draw_line_tests = [
  Alcotest.test_case "draw_line" `Quick (fun () ->
    match line_tool () with
    | None -> Alcotest.skip ()
    | Some tool ->
      let m = empty_layer_model () in
      let (ctx, _) = make_ctx m in
      tool#on_press ctx 10.0 20.0 ~shift:false ~alt:false;
      tool#on_move ctx 30.0 40.0 ~shift:false ~alt:false ~dragging:true;
      tool#on_release ctx 50.0 60.0 ~shift:false ~alt:false;
      let children = layer0_children m in
      assert (Array.length children = 1);
      (match children.(0) with
       | Element.Line l ->
         assert (l.x1 = 10.0);
         assert (l.y1 = 20.0);
         assert (l.x2 = 50.0);
         assert (l.y2 = 60.0)
       | _ -> assert false));
]

(* ── short_line_not_created: zero-length drag commits nothing ── *)

let short_line_tests = [
  Alcotest.test_case "short_line_not_created" `Quick (fun () ->
    match line_tool () with
    | None -> Alcotest.skip ()
    | Some tool ->
      let m = empty_layer_model () in
      let (ctx, _) = make_ctx m in
      (* Press and release at the same point — hypot distance = 0. *)
      tool#on_press ctx 10.0 20.0 ~shift:false ~alt:false;
      tool#on_release ctx 10.0 20.0 ~shift:false ~alt:false;
      assert (Array.length (layer0_children m) = 0));
]

(* ── idle_after_release: mode latches idle -> drawing -> idle ── *)

let idle_after_release_tests = [
  Alcotest.test_case "idle_after_release" `Quick (fun () ->
    match line_tool () with
    | None -> Alcotest.skip ()
    | Some tool ->
      let m = empty_layer_model () in
      let (ctx, _) = make_ctx m in
      assert (mode tool = "idle");
      tool#on_press ctx 10.0 20.0 ~shift:false ~alt:false;
      assert (mode tool = "drawing");
      tool#on_release ctx 50.0 60.0 ~shift:false ~alt:false;
      assert (mode tool = "idle"));
]

(* ── move_without_press_is_noop: guarded by mode == drawing ── *)

let move_noop_tests = [
  Alcotest.test_case "move_without_press_is_noop" `Quick (fun () ->
    match line_tool () with
    | None -> Alcotest.skip ()
    | Some tool ->
      let m = empty_layer_model () in
      let (ctx, _) = make_ctx m in
      (* on_mousemove's handler is guarded by mode == drawing; without a
         prior on_mousedown, mode stays idle and nothing happens. *)
      tool#on_move ctx 50.0 60.0 ~shift:false ~alt:false ~dragging:true;
      assert (mode tool = "idle");
      assert (Array.length (layer0_children m) = 0));
]

let () =
  Alcotest.run "Yaml line tool" [
    "Tool load", load_tests;
    "Draw line", draw_line_tests;
    "Short line not created", short_line_tests;
    "Idle after release", idle_after_release_tests;
    "Move without press noop", move_noop_tests;
  ]

File: jas/tools/yaml_tool_line_test.py
"""Line-tool gesture-seam tests — Python port of the Rust line seam tests in
jas_dioxus/src/tools/yaml_tool.rs (the ``line_parity_*`` family).

These drive the production ``line`` tool loaded from the compiled workspace
bundle through the full gesture pipeline — on_press / on_move / on_release — and
assert the externally-observable outcomes (the committed Line element + the
tool's "mode" state).

The line tool is SIMPLE: a press-drag-release commits a single Line whose
endpoints are the press point (x1,y1) and the release point (x2,y2) in doc
space; under the identity view used here doc coords == screen coords. A drag
whose endpoint-to-endpoint distance (hypot) is shorter than 2px is rejected. The
tool reads NO app-level state, so — unlike the blob-brush seam tests — there is
no app-state seeding/bridge step.

Seam mapping from Rust to this app (Python / PySide6):
    on_press            -> on_press(ctx, x, y, shift, alt)
    on_move(drag)       -> on_move(ctx, x, y, shift, alt, dragging)
    on_release          -> on_release(ctx, x, y, shift, alt)
    tool_state("mode")  -> tool.tool_state("mode")

Committed lines land at ``layers[0].children[0]``.
"""

from __future__ import annotations

import json
import os
import sys

_REPO_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", ".."))
if _REPO_ROOT not in sys.path:
    sys.path.insert(0, _REPO_ROOT)
_JAS_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
if _JAS_DIR not in sys.path:
    sys.path.insert(0, _JAS_DIR)

import pytest

from document.controller import Controller
from document.document import Document
from document.model import Model
from geometry.element import Layer, Line
from tools.tool import ToolContext
from tools.yaml_tool import YamlTool


# ── Loader + harness helpers ───────────────────────────────


def _line_tool() -> YamlTool | None:
    """Load the "line" spec from the compiled workspace bundle and build a
    YamlTool, mirroring the loader in yaml_tool_blob_brush_test.py."""
    ws_path = os.path.abspath(os.path.join(
        _REPO_ROOT, "workspace", "workspace.json",
    ))
    if not os.path.exists(ws_path):
        return None
    with open(ws_path, "r") as f:
        data = json.load(f)
    tools = data.get("tools")
    if not isinstance(tools, dict):
        return None
    spec = tools.get("line")
    return YamlTool.from_workspace_tool(spec) if spec else None


def _empty_layer_model() -> Model:
    """A document with a single empty top-level layer — the empty-layer model
    the Rust port builds (committed lines land at layers[0].children[0])."""
    layer = Layer(name="L", children=())
    return Model(document=Document(layers=(layer,)))


def _ctx(model: Model) -> ToolContext:
    """A faithful headless ToolContext over a live document, with inert
    hit-test callbacks (the line tool does no hit-testing). Mirrors the
    seam-test harness ToolContext."""
    return ToolContext(
        model=model,
        controller=Controller(model=model),
        hit_test_selection=lambda x, y: False,
        hit_test_handle=lambda x, y: None,
        hit_test_text=lambda x, y: None,
        hit_test_path_curve=lambda x, y: None,
        request_update=lambda: None,
    )


def _children(model: Model):
    return model.document.layers[0].children


class TestLineTool:
    # ── Loader sanity (non-vacuity guard) ──────────────────

    def test_line_tool_loads_from_workspace(self):
        tool = _line_tool()
        if tool is None:
            pytest.skip("workspace.json not available")
        assert tool.spec.id == "line"

    # ── 1. Draw line commits exactly one Line element ──────

    def test_draw_line(self):
        tool = _line_tool()
        if tool is None:
            pytest.skip("workspace.json not available")
        model = _empty_layer_model()
        ctx = _ctx(model)
        tool.on_press(ctx, 10.0, 20.0, False, False)
        tool.on_move(ctx, 30.0, 40.0, False, False, True)
        tool.on_release(ctx, 50.0, 60.0, False, False)

        children = _children(model)
        assert len(children) == 1, "draw commits exactly one element"
        le = children[0]
        assert isinstance(le, Line), f"expected Line, got {type(le).__name__}"
        assert le.x1 == 10.0
        assert le.y1 == 20.0
        assert le.x2 == 50.0
        assert le.y2 == 60.0

    # ── 2. Zero-length drag commits nothing ────────────────

    def test_short_line_not_created(self):
        tool = _line_tool()
        if tool is None:
            pytest.skip("workspace.json not available")
        model = _empty_layer_model()
        ctx = _ctx(model)
        # Press and release at same point — hypot distance = 0.
        tool.on_press(ctx, 10.0, 20.0, False, False)
        tool.on_release(ctx, 10.0, 20.0, False, False)
        assert len(_children(model)) == 0, (
            "a drag shorter than 2px is rejected — no line committed")

    # ── 3. Mode latches drawing on press, idle on release ──

    def test_idle_after_release(self):
        tool = _line_tool()
        if tool is None:
            pytest.skip("workspace.json not available")
        model = _empty_layer_model()
        ctx = _ctx(model)
        assert tool.tool_state("mode") == "idle"
        tool.on_press(ctx, 10.0, 20.0, False, False)
        assert tool.tool_state("mode") == "drawing"
        tool.on_release(ctx, 50.0, 60.0, False, False)
        assert tool.tool_state("mode") == "idle"

    # ── 4. Move without a prior press is a no-op ───────────

    def test_move_without_press_is_noop(self):
        tool = _line_tool()
        if tool is None:
            pytest.skip("workspace.json not available")
        model = _empty_layer_model()
        ctx = _ctx(model)
        # on_move's handler is guarded by mode == "drawing"; without a prior
        # on_press, mode stays "idle" and nothing happens.
        tool.on_move(ctx, 50.0, 60.0, False, False, True)
        assert tool.tool_state("mode") == "idle"
        assert len(_children(model)) == 0, "no element committed by a bare move"
